etna/search/views.py

```diff
@@ -1,172 +1,208 @@
+    rendering, then added to the template context as "buckets" by
+    `get_context_data()`
     """
 
-    for key, value in aggregations.items():
-        if buckets := value.get("buckets"):
-            field_name = camelcase_to_underscore(key)
-            if field_name in form.dynamic_choice_fields:
-                form.fields[field_name].update_choices(buckets)
-                form[field_name].more_filter_options_available = bool(
-                    value.get("sum_other_doc_count", 0)
-                )
-
-
-def featured_search(request):
-
-    responses = []
-
-    form = FeaturedSearchForm(request.GET)
-
-    if form.is_valid():
-        q = form.cleaned_data.get("q")
-
-        response = Record.api.client.search_all(
-            q=q,
-            filter_aggregations=[f"group:{bucket.key}" for bucket in FEATURED_BUCKETS],
-            size=3,
-        )
-        responses = response.get("responses", [])
-
-    result_groups = {}
-    for i, bucket in enumerate(FEATURED_BUCKETS):
-        result_groups[bucket.key] = responses[i]
-        result_groups[bucket.key]["bucket"] = bucket
-
-    meta_title = "Search results"
-    query = form.cleaned_data.get("q")
-    if query:
-        meta_title += f' for "{query}"'
-
-    return render(
-        request,
-        "search/featured_search.html",
-        {
-            "meta_title": meta_title,
-            "query": query,
-            "form": form,
-            "result_groups": result_groups,
-        },
-    )
-
-
-def search(request):
-    form = CatalogueSearchForm()
-
-    # Make empty search to fetch aggregations
-    response = Record.api.client.search(
-        template=Template.DETAILS,
-        aggregations=[
-            Aggregation.CATALOGUE_SOURCE,
-            Aggregation.CLOSURE,
-            Aggregation.COLLECTION,
-            Aggregation.LEVEL,
-            Aggregation.TOPIC,
-            # Fetching more groups so that we receive a counts
-            # for any bucket/tab options we might be showing
-            f"{Aggregation.GROUP}:30",
-        ],
-        size=0,
-    )
-    bucket_count_response, _ = response["responses"]
-
-    return render(
-        request,
-        "search/search.html",
-        {
-            "meta_title": "Search the collection",
-            "form": form,
-            "bucket_count_response": bucket_count_response,
-            "buckets": CATALOGUE_BUCKETS,
-        },
-    )
-
-
-def catalogue_search(request):
-
-    per_page = int(request.GET.get("per_page", 20))
-    page_number = int(request.GET.get("page", 1))
-    offset = (page_number - 1) * per_page
+    # The source data for get_buckets()
+    bucket_list: BucketList = None
+    # Can be updated by the view for get_current_bucket_key() to pick up
+    current_bucket_key: str = None
+
+    def extract_group_buckets(
+        self, api_result: Union[None, Dict[str, Any]]
+    ) -> Sequence[Dict[str, Union[str, int]]]:
```

Comment:
Quick Python know-how question. Am I correct to understand this definition to be: 

* Stipulating that `extract_group_buckets` expects an `api_result` argument with a type of [Union](https://docs.python.org/3/library/typing.html#typing.Union). What confuses me a little is the presence of `None` in the Union type - but I'm sure this is my ignorance. 
* Stipulating that it returns a Sequence of Dicts (with the key of each being a string and the value being either a `str` or `int`)

Is that right? It would also be great if you could help me understand the `Union[None, Dict[str, Any]]`

Reply:
`typing.Union` essentially means "any one of the enclosed types". So, in this case, the value is expected to be `None` in some circumstances, and a dictionary in others. However, it always must always be provided.